gm4_cement_mixers/data/cement_mixers/functions/item_fill/black_concrete.mcfunction

```diff
@@ -0,0 +1,3 @@
+replaceitem block ~ ~ ~ container.0 black_concrete 1
+scoreboard players remove @s gm4_lt_buffer 1
+tag @s add gm4_lt_fill
```

Comment:
`tag @s add gm4_lt_fill`

Concrete can't be drained. There is no need for this tag.
I even think we should add a separate hook for filling(/draining) items where their counterpart can't be drained(/filled) again. The current system would need a module tick where the tank has no items for it to be able to make concrete again

Reply:
The `gm4_lt_fill` tag informs the tank that the last action it performed was filling an item with liquid, preventing the next tick from attempting to drain that same item. Even if concrete cant be drained, the tag lowers the number of checks the tank performs; it dosen't go through the list of drainable items at all. This partially increases performance, even though it isn't needed to prevent a loop.
The current system doesn't need a tick with an empty hopper to make concrete; The tank can do the same operation (filling or draining) in consecutive module pulses.

Reply:
Mmh, I thought the tag was only removed if there is no item is slot 0 and this module places the Concrete there.

Reply:
The tag Is only removed if the first slot is empty. That does not prevent the module from placing another concrete (filling operation) though, only from attempting to drain the item (as if it were a bucket) the next clock pulse around.

Reply:
That's the thing. If you want to automate this you have to fill the concrete, wait a module tick, and only then you can fill it with water again.

Reply:
He's right. Just tested it and swapping an empty just-drained bucket with powder causes the tank to do nothing.

Reply:
Actually just removed the tag from all of the fill functions and the tank still doesn't respond to the powder until the tank detects the slot is empty.

Reply:
> Actually just removed the tag from all of the fill functions and the tank still doesn't respond to the powder until the tank detects the slot is empty.

Yes, this is because the standard liquids adds the tag. That's why I made this comment.
> I even think we should add a separate hook for filling(/draining) items where their counterpart can't be drained(/filled) again. The current system would need a module tick where the tank has no items for it to be able to make concrete again

The other way around (water after concrete) should work with the tag removed.